subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/DefaultVirtualFileSystem.java

```diff
@@ -207,4 +211,13 @@ public void updateWithKnownSnapshot(CompleteFileSystemLocationSnapshot snapshot)
             }
         }
     }
+
+    public synchronized void updateDefaultExcludes(String... newDefaultExcludesArgs) {
```

Comment:
I don't like this synchronization here. This is supposed to be called only when `settingsEvaluated()`, right? Do we need to synchronize at all?

I'm also wondering if we could avoid the mutable state altogether. Could we create some new object with the modified `DirectorySnapshotter` for the part of the build that comes after evaluating settings somehow?

Reply:
I removed the synchronization. As you said, the `settingsEvaluated` hook should not happen concurrently.

Reply:
The virtual file system is Gradle user home scoped. It seems like it needs to be mutable if it should be able to snapshot things with the new default excludes.